### ISSUE TITLE
Out of direct memory in the Engine

### ISSUE THREAD

**Issue by Zelldon** [zeebe-io/zeebe#3686]
**Description**

![died](https://user-images.githubusercontent.com/2758593/72519634-0a576a80-3858-11ea-8e3d-73eaabe391e4.png)

First look at this metrics it looked like that back pressure never releases the requests correctly.
But it seem that the real problem is that the engine uses to much direct memory.

![died-res](https://user-images.githubusercontent.com/2758593/72519688-2b1fc000-3858-11ea-826d-f05de06110fc.png)

```
2020-01-16T10:41:09.659753939Z 2020-01-16 10:41:09.659 [Broker-0-StreamProcessor-2] [Broker-0-zb-actors-2] ERROR io.zeebe.util.actor - Actor failed in phase 'STARTED'. Continue with next job.
 I 
2020-01-16T10:41:09.659787894Z java.lang.OutOfMemoryError: Direct buffer memory
	at java.nio.Bits.reserveMemory(Unknown Source) ~[?:?]
	at java.nio.DirectByteBuffer.<init>(Unknown Source) ~[?:?]
	at java.nio.ByteBuffer.allocateDirect(Unknown Source) ~[?:?]
	at org.agrona.ExpandableDirectByteBuffer.ensureCapacity(ExpandableDirectByteBuffer.java:1128) ~[agrona-1.1.0.jar:1.1.0]
	at org.agrona.ExpandableDirectByteBuffer.putLong(ExpandableDirectByteBuffer.java:184) ~[agrona-1.1.0.jar:1.1.0]
	at io.zeebe.logstreams.impl.log.LogStreamBatchWriterImpl.copyExistingEventToBuffer(LogStreamBatchWriterImpl.java:169) ~[zeebe-logstreams-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.logstreams.impl.log.LogStreamBatchWriterImpl.event(LogStreamBatchWriterImpl.java:82) ~[zeebe-logstreams-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.processor.TypedCommandWriterImpl.appendRecord(TypedCommandWriterImpl.java:73) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.processor.TypedCommandWriterImpl.appendRecord(TypedCommandWriterImpl.java:62) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.processor.TypedStreamWriterImpl.appendFollowUpEvent(TypedStreamWriterImpl.java:63) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.processor.workflow.UpdateVariableStreamWriter.onCreate(UpdateVariableStreamWriter.java:41) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.state.instance.VariablesState.setVariableLocal(VariablesState.java:167) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.state.instance.VariablesState.setVariablesLocalFromDocument(VariablesState.java:114) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.processor.workflow.instance.CreateWorkflowInstanceProcessor.setVariablesFromDocument(CreateWorkflowInstanceProcessor.java:114) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.processor.workflow.instance.CreateWorkflowInstanceProcessor.onCommand(CreateWorkflowInstanceProcessor.java:78) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.processor.CommandProcessorImpl.processRecord(CommandProcessorImpl.java:47) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.processor.TypedRecordProcessor.processRecord(TypedRecordProcessor.java:28) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.processor.TypedRecordProcessor.processRecord(TypedRecordProcessor.java:48) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.processor.ReProcessingStateMachine.lambda$chooseOperationForEvent$5(ReProcessingStateMachine.java:277) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransaction.run(ZeebeTransaction.java:79) ~[zeebe-db-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.processor.ReProcessingStateMachine.lambda$processUntilDone$2(ReProcessingStateMachine.java:252) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.retry.ActorRetryMechanism.run(ActorRetryMechanism.java:36) ~[zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.retry.EndlessRetryStrategy.run(EndlessRetryStrategy.java:46) ~[zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:73) ~[zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:39) [zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:115) [zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:107) [zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:91) [zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:195) [zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
 I 
2020-01-16T10:41:09.660027242Z 2020-01-16 10:41:09.659 [Broker-0-StreamProcessor-1] [Broker-0-zb-actors-1] ERROR io.zeebe.util.actor - Actor failed in phase 'STARTED'. Continue with next job.
 I 
2020-01-16T10:41:09.660031030Z java.lang.OutOfMemoryError: Direct buffer memory
	at java.nio.Bits.reserveMemory(Unknown Source) ~[?:?]
	at java.nio.DirectByteBuffer.<init>(Unknown Source) ~[?:?]
	at java.nio.ByteBuffer.allocateDirect(Unknown Source) ~[?:?]
	at org.agrona.ExpandableDirectByteBuffer.ensureCapacity(ExpandableDirectByteBuffer.java:1128) ~[agrona-1.1.0.jar:1.1.0]
	at org.agrona.ExpandableDirectByteBuffer.putLong(ExpandableDirectByteBuffer.java:184) ~[agrona-1.1.0.jar:1.1.0]
	at io.zeebe.logstreams.impl.log.LogStreamBatchWriterImpl.copyExistingEventToBuffer(LogStreamBatchWriterImpl.java:169) ~[zeebe-logstreams-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.logstreams.impl.log.LogStreamBatchWriterImpl.event(LogStreamBatchWriterImpl.java:82) ~[zeebe-logstreams-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.processor.TypedCommandWriterImpl.appendRecord(TypedCommandWriterImpl.java:73) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.processor.TypedCommandWriterImpl.appendRecord(TypedCommandWriterImpl.java:62) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.processor.TypedStreamWriterImpl.appendFollowUpEvent(TypedStreamWriterImpl.java:63) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.processor.workflow.UpdateVariableStreamWriter.onCreate(UpdateVariableStreamWriter.java:41) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.state.instance.VariablesState.setVariableLocal(VariablesState.java:167) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.state.instance.VariablesState.setVariablesLocalFromDocument(VariablesState.java:114) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.processor.workflow.instance.CreateWorkflowInstanceProcessor.setVariablesFromDocument(CreateWorkflowInstanceProcessor.java:114) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.processor.workflow.instance.CreateWorkflowInstanceProcessor.onCommand(CreateWorkflowInstanceProcessor.java:78) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.processor.CommandProcessorImpl.processRecord(CommandProcessorImpl.java:47) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.processor.TypedRecordProcessor.processRecord(TypedRecordProcessor.java:28) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.processor.TypedRecordProcessor.processRecord(TypedRecordProcessor.java:48) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.processor.ReProcessingStateMachine.lambda$chooseOperationForEvent$5(ReProcessingStateMachine.java:277) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransaction.run(ZeebeTransaction.java:79) ~[zeebe-db-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.processor.ReProcessingStateMachine.lambda$processUntilDone$2(ReProcessingStateMachine.java:252) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.retry.ActorRetryMechanism.run(ActorRetryMechanism.java:36) ~[zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.retry.EndlessRetryStrategy.run(EndlessRetryStrategy.java:46) ~[zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:73) ~[zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:39) [zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:115) [zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:107) [zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:91) [zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:195) [zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
 I 
2020-01-16T10:41:09.746119850Z 2020-01-16 10:41:09.745 [Broker-0-StreamProcessor-3] [Broker-0-zb-actors-3] ERROR io.zeebe.util.actor - Actor failed in phase 'STARTED'. Continue with next job.
 I 
2020-01-16T10:41:09.746142302Z java.lang.OutOfMemoryError: Direct buffer memory
	at java.nio.Bits.reserveMemory(Unknown Source) ~[?:?]
	at java.nio.DirectByteBuffer.<init>(Unknown Source) ~[?:?]
	at java.nio.ByteBuffer.allocateDirect(Unknown Source) ~[?:?]
	at org.agrona.ExpandableDirectByteBuffer.ensureCapacity(ExpandableDirectByteBuffer.java:1128) ~[agrona-1.1.0.jar:1.1.0]
	at org.agrona.ExpandableDirectByteBuffer.putLong(ExpandableDirectByteBuffer.java:184) ~[agrona-1.1.0.jar:1.1.0]
	at io.zeebe.logstreams.impl.log.LogStreamBatchWriterImpl.copyExistingEventToBuffer(LogStreamBatchWriterImpl.java:169) ~[zeebe-logstreams-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.logstreams.impl.log.LogStreamBatchWriterImpl.event(LogStreamBatchWriterImpl.java:82) ~[zeebe-logstreams-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.processor.TypedCommandWriterImpl.appendRecord(TypedCommandWriterImpl.java:73) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.processor.TypedCommandWriterImpl.appendRecord(TypedCommandWriterImpl.java:62) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.processor.TypedStreamWriterImpl.appendFollowUpEvent(TypedStreamWriterImpl.java:63) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.processor.workflow.UpdateVariableStreamWriter.onCreate(UpdateVariableStreamWriter.java:41) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.state.instance.VariablesState.setVariableLocal(VariablesState.java:167) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.state.instance.VariablesState.setVariablesLocalFromDocument(VariablesState.java:114) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.processor.workflow.instance.CreateWorkflowInstanceProcessor.setVariablesFromDocument(CreateWorkflowInstanceProcessor.java:114) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.processor.workflow.instance.CreateWorkflowInstanceProcessor.onCommand(CreateWorkflowInstanceProcessor.java:78) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.processor.CommandProcessorImpl.processRecord(CommandProcessorImpl.java:47) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.processor.TypedRecordProcessor.processRecord(TypedRecordProcessor.java:28) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.processor.TypedRecordProcessor.processRecord(TypedRecordProcessor.java:48) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.processor.ReProcessingStateMachine.lambda$chooseOperationForEvent$5(ReProcessingStateMachine.java:277) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransaction.run(ZeebeTransaction.java:79) ~[zeebe-db-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.engine.processor.ReProcessingStateMachine.lambda$processUntilDone$2(ReProcessingStateMachine.java:252) ~[zeebe-workflow-engine-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.retry.ActorRetryMechanism.run(ActorRetryMechanism.java:36) ~[zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.retry.EndlessRetryStrategy.run(EndlessRetryStrategy.java:46) ~[zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:73) ~[zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:39) [zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:115) [zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:107) [zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:91) [zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:195) [zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
 I 
2020-01-16T10:41:10.027598512Z Jan 16, 2020 10:41:10 AM org.elasticsearch.client.RestClient logResponse
 E 
```

Locally on one of my machines I can't run the engine tests, because it will go out of direct memory as well, might be a hint here as well...

_**Note:**_
Actually usage of direct memory in the engine is no longer necessary, since we migrated the log completely to atomix, which depends only on heap (byte arrays).

**Comment by Zelldon** [comment]
After creating an heap dump on `zeebe-0` and searching for the ExpandableDirectBuffer I probably found the issue.

![expandable-buffer](https://user-images.githubusercontent.com/2758593/72528235-e2253700-386a-11ea-9dd3-5ce2e291d7fa.png)

We have this buffer in each LogStreamBatchWriter, which is bound to a partition. 
Each buffer holds 827,430,261 bytes, which are ~ 800 Mb, which means we have ~2,4 gig direct memory with these writers.

**Comment by Zelldon** [comment]
@npepinpe you asked recently for ExpandableDirectBuffers :sweat_smile: :laughing:

**Comment by npepinpe** [comment]
https://sadtrombone.com/ :(

Does this imply that we sometimes have 800 Mb as a single batch? :scream: I assume we always restart at 0 between batches?

**Comment by Zelldon** [comment]
Seems that the reset does not work properly

**Comment by Zelldon** [comment]
Seems to be related to Reprocessing, because it always happens after I kill a pod and I added the following to the `LogStreamBatchWriterImpl#copyExistingEventToBuffer`

```

    if (eventBufferOffset > 32 * 1024 * 1024) {
      LoggerFactory.getLogger(this.getClass().getName())
          .error(
              "Something happens, we have a big eventBufferOffset {} and lot events to append {}, last event length {}",
              eventBufferOffset,
              eventCount,
              eventLength,
              new IllegalStateException());
    }
```

I run a benchmark with that and killed a pod. After leader change this was printed:

```
I 2020-01-16T14:21:14.710805107Z 2020-01-16 14:21:14.710 [Broker-0-StreamProcessor-1] [Broker-1-zb-actors-2] INFO  io.zeebe.processor - Processor starts reprocessing, until last source event position 992141608576
 
I 2020-01-16T14:21:17.250233520Z 2020-01-16 14:21:17.249 [Broker-0-StreamProcessor-2] [Broker-1-zb-actors-3] INFO  io.zeebe.processor - Processor starts reprocessing, until last source event position 1039389022480
 
I 2020-01-16T14:21:25.447401944Z 2020-01-16 14:21:25.446 [Broker-0-StreamProcessor-2] [Broker-1-zb-actors-2] ERROR io.zeebe.logstreams.impl.log.LogStreamBatchWriterImpl - Something happens, we have a big eventBufferOffset 33561688 and lot events to append 9350, last event length 33374688
 
I 2020-01-16T14:21:25.447931541Z 2020-01-16 14:21:25.447 [Broker-0-StreamProcessor-2] [Broker-1-zb-actors-2] ERROR io.zeebe.logstreams.impl.log.LogStreamBatchWriterImpl - Something happens, we have a big eventBufferOffset 33561839 and lot events to append 9351, last event length 33374819
 
I 2020-01-16T14:21:25.448361851Z 2020-01-16 14:21:25.447 [Broker-0-StreamProcessor-2] [Broker-1-zb-actors-2] ERROR io.zeebe.logstreams.impl.log.LogStreamBatchWriterImpl - Something happens, we have a big eventBufferOffset 33561990 and lot events to append 9352, last event length 33374950
 
I 2020-01-16T14:21:25.451174928Z 2020-01-16 14:21:25.450 [Broker-0-StreamProcessor-2] [Broker-1-zb-actors-2] ERROR io.zeebe.logstreams.impl.log.LogStreamBatchWriterImpl - Something happens, we have a big eventBufferOffset 33562141 and lot events to append 9353, last event length 33375081
 
I 2020-01-16T14:21:25.451742750Z 2020-01-16 14:21:25.451 [Broker-0-StreamProcessor-2] [Broker-1-zb-actors-2] ERROR io.zeebe.logstreams.impl.log.LogStreamBatchWriterImpl - Something happens, we have a big eventBufferOffset 33576043 and lot events to append 9354, last event length 33388963
 ```

It stops with:

```
I 2020-01-16T14:23:34.211810358Z 2020-01-16 14:23:34.211 [Broker-0-StreamProcessor-2] [Broker-1-zb-actors-0] ERROR io.zeebe.logstreams.impl.log.LogStreamBatchWriterImpl - Something happens, we have a big eventBufferOffset 275271329 and lot events to append 76703, last event length 273737269
 
I 2020-01-16T14:23:34.212053297Z 2020-01-16 14:23:34.211 [Broker-0-StreamProcessor-2] [Broker-1-zb-actors-0] ERROR io.zeebe.logstreams.impl.log.LogStreamBatchWriterImpl - Something happens, we have a big eventBufferOffset 275271480 and lot events to append 76704, last event length 273737400
 
I 2020-01-16T14:23:34.232993732Z 2020-01-16 14:23:34.232 [Broker-0-StreamProcessor-2] [Broker-1-zb-actors-3] INFO  io.zeebe.processor - Processor finished reprocessing at event position 1039389022480
```

So might be that we are reseting the writer here not correctly, since we not writing at all on reprocessing?

**Comment by npepinpe** [comment]
@menski and I came up with a different fix, I think it might help. Essentially we make sure to not write on repro. PR incoming

**Comment by menski** [comment]
With the fix from #3693 after multiple restarts the following buffer capacities where observed

### zeebe-0
![ExpandableDirectByteBuffer-zeebe-0](https://user-images.githubusercontent.com/516296/72549636-596dc180-3891-11ea-80fe-ba72c05b0dc4.png)

### zeebe-1

Is not leader of any partition and therefore has no `ExpandableDirectByteBuffer`

### zeebe-2
![ExpandableDirectByteBuffer-zeebe-2](https://user-images.githubusercontent.com/516296/72549651-62f72980-3891-11ea-8404-971470a03d77.png)
